src/backend/distributed/planner/multi_router_planner.c

```diff
@@ -2066,41 +2106,162 @@ TargetShardIntervalForModify(Query *query)
 								"and try again.")));
 	}
 
+	if (commandType == CMD_INSERT)
+	{
+		return FindShardForInsert(query);
+	}
+	else
+	{
+		List *restrictClauseList = QueryRestrictList(query);
+		Index tableId = 1;
+		List *shardIntervalList = NIL;
+
+		shardIntervalList = LoadShardIntervalList(distributedTableId);
+		prunedShardList = PruneShardList(distributedTableId, tableId, restrictClauseList,
+										 shardIntervalList);
+	}
+
+	prunedShardCount = list_length(prunedShardList);
+	if (prunedShardCount != 1)
+	{
+		Oid relationId = cacheEntry->relationId;
+		char *partitionKeyString = cacheEntry->partitionKeyString;
+		char *partitionColumnName = ColumnNameToColumn(relationId, partitionKeyString);
+		StringInfo errorHint = makeStringInfo();
+		const char *targetCountType = NULL;
+		bool showHint = false;
+
+		if (prunedShardCount == 0)
+		{
+			targetCountType = "no";
+		}
+		else
+		{
+			targetCountType = "multiple";
+		}
+
+		appendStringInfo(errorHint, "Consider using an equality filter on "
+									"partition column \"%s\" to target a "
+									"single shard. If you'd like to run a "
+									"multi-shard operation, use "
+									"master_modify_multiple_shards().",
+						 partitionColumnName);
+
+		if (commandType == CMD_DELETE && partitionMethod == DISTRIBUTE_BY_APPEND)
+		{
+			appendStringInfo(errorHint, " You can also use "
+										"master_apply_delete_command() to drop "
+										"all shards satisfying delete criteria.");
+		}
+
+		showHint = errorHint->len > 0;
+
+		ereport(ERROR, (errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
+						errmsg("cannot run %s command which targets %s shards",
+							   commandName, targetCountType),
+						showHint ? errhint("%s", errorHint->data) : 0));
+	}
+
+	return (ShardInterval *) linitial(prunedShardList);
+}
+
+
+/*
+ * FindShardForInsert finds the shard interval for an INSERT query or
+ * NULL if the partition column value is defined as an expression that
+ * still needs to be evaluated.
+ */
+ShardInterval *
+FindShardForInsert(Query *query)
+{
+	Oid distributedTableId = ExtractFirstDistributedTableId(query);
+	DistTableCacheEntry *cacheEntry = DistributedTableCacheEntry(distributedTableId);
```

Comment:
Since this is a public function that is also called from the executor, it seems slightly nicer to avoid putting the burden on the caller here.